File: flutter_appauth/ios/Classes/FlutterAppAuth.m
#import "FlutterAppAuth.h"

@implementation FlutterAppAuth

+ (NSMutableDictionary *)processResponses:(OIDTokenResponse*) tokenResponse authResponse:(OIDAuthorizationResponse*) authResponse {
    NSMutableDictionary *processedResponses = [[NSMutableDictionary alloc] init];
    if(tokenResponse.accessToken) {
        [processedResponses setValue:tokenResponse.accessToken forKey:@"accessToken"];
    }
    if(tokenResponse.accessTokenExpirationDate) {
        [processedResponses setValue:[[NSNumber alloc] initWithDouble:[tokenResponse.accessTokenExpirationDate timeIntervalSince1970] * 1000] forKey:@"accessTokenExpirationTime"];
    }
    if(authResponse) {
        if (authResponse.additionalParameters) {
            [processedResponses setObject:authResponse.additionalParameters forKey:@"authorizationAdditionalParameters"];
        }
        if (authResponse.request && authResponse.request.nonce) {
            [processedResponses setObject:authResponse.request.nonce forKey:@"nonce"];
        }
     }
    if(tokenResponse.additionalParameters) {
        [processedResponses setObject:tokenResponse.additionalParameters forKey:@"tokenAdditionalParameters"];
    }
    if(tokenResponse.idToken) {
        [processedResponses setValue:tokenResponse.idToken forKey:@"idToken"];
    }
    if(tokenResponse.refreshToken) {
        [processedResponses setValue:tokenResponse.refreshToken forKey:@"refreshToken"];
    }
    if(tokenResponse.tokenType) {
        [processedResponses setValue:tokenResponse.tokenType forKey:@"tokenType"];
    }
    if (tokenResponse.scope) {
        [processedResponses setObject:[tokenResponse.scope componentsSeparatedByString: @" "] forKey:@"scopes"];
    }
    
    return processedResponses;
}

+ (void)finishWithError:(NSString *)errorCode message:(NSString *)message result:(FlutterResult)result error:(NSError * _Nullable)error {

    NSMutableDictionary<NSString *, id> *details = [NSMutableDictionary dictionary];

    if (error) {
        id authError = error.userInfo[OIDOAuthErrorResponseErrorKey];
        NSDictionary<NSString *, id> *authErrorMap = [authError isKindOfClass:[NSDictionary class]] ? authError : nil;
        
        if (authErrorMap) {
            if ([authErrorMap objectForKey:OIDOAuthErrorFieldError]) {
                [details setObject:authErrorMap[OIDOAuthErrorFieldError] forKey:OIDOAuthErrorFieldError];
            }
            if ([authErrorMap objectForKey:OIDOAuthErrorFieldErrorDescription]) {
                [details setObject:authErrorMap[OIDOAuthErrorFieldErrorDescription] forKey:OIDOAuthErrorFieldErrorDescription];
            }
            if ([authErrorMap objectForKey:OIDOAuthErrorFieldErrorURI]) {
                [details setObject:authErrorMap[OIDOAuthErrorFieldErrorURI] forKey:OIDOAuthErrorFieldErrorURI];
            }
        }
        if (error.domain) {
            [details setObject:error.domain forKey:@"type"];
        }
        if (error.code) {
            [details setObject:[@(error.code) stringValue] forKey:@"code"];
        }
        
        id underlyingErr = [error.userInfo objectForKey:NSUnderlyingErrorKey];
        NSError *underlyingError = [underlyingErr isKindOfClass:[NSError class]] ? underlyingErr : nil;
        if (underlyingError) {
            if (underlyingError.domain) {
                [details setObject:underlyingError.domain forKey:@"domain"];
            }

            if (underlyingError.debugDescription) {
                [details setObject:underlyingError.debugDescription forKey:@"root_cause_debug_description"];
            }
        }
        
        if (error.debugDescription) {
            [details setObject:error.debugDescription forKey:@"error_debug_description"];
        }
        
        bool userDidCancel = [error.domain  isEqual: @"org.openid.appauth.general"] 
                             && error.code == OIDErrorCodeUserCanceledAuthorizationFlow;
        [details setObject:(userDidCancel ? @"true" : @"false") forKey:@"user_did_cancel"];

    }
    result([FlutterError errorWithCode:errorCode message:message details:details]);
}

+ (NSString *) formatMessageWithError:(NSString *)messageFormat error:(NSError * _Nullable)error {
    NSString *formattedMessage = [NSString stringWithFormat:messageFormat, [error localizedDescription]];
    return formattedMessage;
}

@end

@implementation AppAuthAuthorization

- (id<OIDExternalUserAgentSession>)performAuthorization:(OIDServiceConfiguration *)serviceConfiguration clientId:(NSString*)clientId clientSecret:(NSString*)clientSecret scopes:(NSArray *)scopes redirectUrl:(NSString*)redirectUrl additionalParameters:(NSDictionary *)additionalParameters externalUserAgent:(NSString*)externalUserAgent result:(FlutterResult)result exchangeCode:(BOOL)exchangeCode nonce:(NSString*)nonce {
    return nil;
}

- (id<OIDExternalUserAgentSession>)performEndSessionRequest:(OIDServiceConfiguration *)serviceConfiguration requestParameters:(EndSessionRequestParameters *)requestParameters result:(FlutterResult)result {
    return nil;
}

@end

File: flutter_appauth/ios/Classes/FlutterAppAuth.swift
#if os(macOS)
import FlutterMacOS
#else
import Flutter
#endif

import AppAuth
import Foundation

class FlutterAppAuth: NSObject {
    static func processResponses(_ tokenResponse: OIDTokenResponse, authResponse: OIDAuthorizationResponse?) -> [String: Any] {
        var result = [String: Any]()
        if let tokenResponse = tokenResponse {
            result["accessToken"] = tokenResponse.accessToken
            result["idToken"] = tokenResponse.idToken
            result["refreshToken"] = tokenResponse.refreshToken
            result["tokenType"] = tokenResponse.tokenType
            result["expiresIn"] = tokenResponse.expiresIn
        }
        if let authResponse = authResponse {
            result["authorizationCode"] = authResponse.authorizationCode
            result["state"] = authResponse.state
        }
        return result
    }

    static func finishWithError(_ errorCode: String, message: String, result: FlutterResult) {
        result(FlutterError(code: errorCode, message: message, details: nil))
    }

    static func formatMessage(with error: Error?) -> String {
        if let error = error {
            return String(format: error.localizedDescription)
        } else {
            return ""
        }
    }
}

let AUTHORIZE_METHOD = "authorize"
let AUTHORIZE_AND_EXCHANGE_CODE_METHOD = "authorizeAndExchangeCode"
let TOKEN_METHOD = "token"
let END_SESSION_METHOD = "endSession"
let AUTHORIZE_ERROR_CODE = "authorize_failed"
let AUTHORIZE_AND_EXCHANGE_CODE_ERROR_CODE = "authorize_and_exchange_code_failed"
let DISCOVERY_ERROR_CODE = "discovery_failed"
let TOKEN_ERROR_CODE = "token_failed"
let END_SESSION_ERROR_CODE = "end_session_failed"
let DISCOVERY_ERROR_MESSAGE_FORMAT = "Error retrieving discovery document: %@"
let TOKEN_ERROR_MESSAGE_FORMAT = "Failed to get token: %@"
let AUTHORIZE_ERROR_MESSAGE_FORMAT = "Failed to authorize: %@"
let END_SESSION_ERROR_MESSAGE_FORMAT = "Failed to end session: %@"

struct EndSessionRequestParameters {
    var idTokenHint: String?
    var postLogoutRedirectUrl: String?
    var state: String?
    var issuer: String?
    var discoveryUrl: String?
    var serviceConfigurationParameters: [String: Any]?
    var additionalParameters: [String: Any]?
    var preferEphemeralSession: Bool
}

class AppAuthAuthorization: NSObject {
    func performAuthorization(_ serviceConfiguration: OIDServiceConfiguration, clientId: String, clientSecret: String, scopes: [String], redirectUrl: String, additionalParameters: [String: Any], preferEphemeralSession: Bool, result: FlutterResult, exchangeCode: Bool, nonce: String?) -> OIDExternalUserAgentSession? {
        let request = OIDAuthorizationRequest(
            configuration: serviceConfiguration,
            clientId: clientId,
            clientSecret: clientSecret,
            scopes: scopes,
            redirectURL: URL(string: redirectUrl)!,
            responseType: exchangeCode ? OIDResponseTypeCode : OIDResponseTypeToken,
            additionalParameters: additionalParameters
        )

        if let nonce = nonce {
            request.nonce = nonce
        }

        let session = OIDAuthState.authState(byPresenting: request, externalUserAgent: OIDExternalUserAgentIOS(presenting: UIApplication.shared.delegate?.window??.rootViewController)) { authorizationResponse, error in
            if let authorizationResponse = authorizationResponse {
                if exchangeCode {
                    self.performTokenExchange(with: authorizationResponse, serviceConfiguration: serviceConfiguration, clientId: clientId, clientSecret: clientSecret, result: result)
                } else {
                    result(FlutterAppAuth.processResponses(nil, authResponse: authorizationResponse))
                }
            } else if let error = error {
                FlutterAppAuth.finishWithError(AUTHORIZE_ERROR_CODE, message: FlutterAppAuth.formatMessage(with: error), result: result)
            }
        }

        return session
    }

    func performTokenExchange(with authorizationResponse: OIDAuthorizationResponse, serviceConfiguration: OIDServiceConfiguration, clientId: String, clientSecret: String, result: FlutterResult) {
        let tokenRequest = OIDTokenRequest(
            configuration: serviceConfiguration,
            grantType: OIDGrantTypeAuthorizationCode,
            authorizationCode: authorizationResponse.authorizationCode,
            redirectURL: authorizationResponse.request.redirectURL,
            clientID: clientId,
            clientSecret: clientSecret,
            scope: authorizationResponse.request.scope,
            refreshToken: nil,
            codeVerifier: authorizationResponse.request.codeVerifier
        )

        OIDAuthorizationService.perform(tokenRequest) { tokenResponse, error in
            if let tokenResponse = tokenResponse {
                result(FlutterAppAuth.processResponses(tokenResponse, authResponse: authorizationResponse))
            } else if let error = error {
                FlutterAppAuth.finishWithError(AUTHORIZE_AND_EXCHANGE_CODE_ERROR_CODE, message: FlutterAppAuth.formatMessage(with: error), result: result)
            }
        }
    }

    func performEndSessionRequest(_ serviceConfiguration: OIDServiceConfiguration, requestParameters: EndSessionRequestParameters, result: FlutterResult) -> OIDExternalUserAgentSession? {
        let endSessionRequest = OIDEndSessionRequest(
            configuration: serviceConfiguration,
            idTokenHint: requestParameters.idTokenHint,
            postLogoutRedirectURL: requestParameters.postLogoutRedirectUrl.flatMap { URL(string: $0) },
            state: requestParameters.state,
            additionalParameters: requestParameters.additionalParameters
        )

        let session = OIDExternalUserAgentSession.present(endSessionRequest, externalUserAgent: OIDExternalUserAgentIOS(presenting: UIApplication.shared.delegate?.window??.rootViewController)) { response, error in
            if let response = response {
                result(["state": response.state])
            } else if let error = error {
                FlutterAppAuth.finishWithError(END_SESSION_ERROR_CODE, message: FlutterAppAuth.formatMessage(with: error), result: result)
            }
        }

        return session
    }
}

